File: backend/internal/service/public_config.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"sort"
	"time"

	"gorm.io/gorm"

	"personal-blog/backend/internal/model"
)

type PublicConfigService struct {
	database       *gorm.DB
	articleService *ArticleService
}

type PublicSiteConfig struct {
	Site        PublicSite             `json:"site"`
	Navigation  []PublicNavigationItem `json:"navigation"`
	SocialLinks []PublicSocialLink     `json:"social_links"`
	Footer      PublicFooter           `json:"footer"`
}

type PublicSite struct {
	Name                 string `json:"name"`
	ShortName            string `json:"short_name"`
	Description          string `json:"description"`
	TitleSuffix          string `json:"title_suffix"`
	LogoURL              string `json:"logo_url"`
	FaviconURL           string `json:"favicon_url"`
	DefaultShareImageURL string `json:"default_share_image_url"`
}

type PublicNavigationItem struct {
	Name         string `json:"name"`
	URL          string `json:"url"`
	LinkType     string `json:"link_type"`
	OpenInNewTab bool   `json:"open_in_new_tab"`
}

type PublicSocialLink struct {
	Platform    string `json:"platform"`
	DisplayName string `json:"display_name"`
	URL         string `json:"url"`
}

type PublicFooter struct {
	CopyrightName  string `json:"copyright_name"`
	StartYear      *int   `json:"start_year"`
	AdditionalText string `json:"additional_text"`
	FilingNumber   string `json:"filing_number"`
	FilingURL      string `json:"filing_url"`
	ShowTechnology bool   `json:"show_technology"`
	TechnologyText string `json:"technology_text"`
}

type PublicHomepage struct {
	Version uint64                 `json:"version"`
	Modules []PublicHomepageModule `json:"modules"`
}

type PublicHomepageModule struct {
	Type             string                              `json:"type"`
	SortOrder        int                                 `json:"sort_order"`
	Hero             *model.HeroModuleConfig             `json:"-"`
	About            *model.AboutModuleConfig            `json:"-"`
	FeaturedArticles *PublicFeaturedArticlesModuleConfig `json:"-"`
	LatestArticles   *PublicLatestArticlesModuleConfig   `json:"-"`
	TechStack        *PublicTechStackModuleConfig        `json:"-"`
	SocialLinks      *PublicSocialLinksModuleConfig      `json:"-"`
}

type PublicArticleSummary struct {
	ID           uint64     `json:"id"`
	Title        string     `json:"title"`
	Slug         string     `json:"slug"`
	Summary      string     `json:"summary"`
	PublishedAt  *time.Time `json:"published_at"`
	CommentCount int64      `json:"comment_count"`
}

type PublicFeaturedArticlesModuleConfig struct {
	Title       string                 `json:"title"`
	Description string                 `json:"description"`
	Limit       int                    `json:"limit"`
	Articles    []PublicArticleSummary `json:"articles"`
}

type PublicLatestArticlesModuleConfig struct {
	Title            string                 `json:"title"`
	Description      string                 `json:"description"`
	Limit            int                    `json:"limit"`
	ShowSummary      bool                   `json:"show_summary"`
	ShowDate         bool                   `json:"show_date"`
	ShowCommentCount bool                   `json:"show_comment_count"`
	ShowViewAll      bool                   `json:"show_view_all"`
	Articles         []PublicArticleSummary `json:"articles"`
}

type PublicTechStackModuleConfig struct {
	Title       string           `json:"title"`
	Description string           `json:"description"`
	Items       []PublicTechItem `json:"items"`
}

type PublicTechItem struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	IconURL     string `json:"icon_url"`
	URL         string `json:"url"`
}

type PublicSocialLinksModuleConfig struct {
	Title       string             `json:"title"`
	Description string             `json:"description"`
	Links       []PublicSocialLink `json:"links"`
}

func NewPublicConfigService(database *gorm.DB) *PublicConfigService {
	return &PublicConfigService{
		database:       database,
		articleService: NewArticleService(database),
	}
}

func (service *PublicConfigService) GetSiteConfig(ctx context.Context) (PublicSiteConfig, error) {
	setting, err := service.siteSetting(ctx)
	if err != nil {
		return PublicSiteConfig{}, err
	}

	var navigationItems []model.NavigationItem
	if err := service.database.WithContext(ctx).
		Where("is_visible = ?", true).
		Order("sort_order ASC, id ASC").
		Find(&navigationItems).Error; err != nil {
		return PublicSiteConfig{}, fmt.Errorf("list public navigation: %w", err)
	}

	var socialLinks []model.SocialLink
	if err := service.database.WithContext(ctx).
		Where("is_visible = ?", true).
		Order("sort_order ASC, id ASC").
		Find(&socialLinks).Error; err != nil {
		return PublicSiteConfig{}, fmt.Errorf("list public social links: %w", err)
	}

	return publicSiteConfig(setting, validPublicNavigation(navigationItems), validPublicSocialLinks(socialLinks)), nil
}

func (service *PublicConfigService) GetHomepage(ctx context.Context) (PublicHomepage, error) {
	version, config, err := service.publishedHomepageConfig(ctx)
	if err != nil {
		return PublicHomepage{}, err
	}
	return service.composeHomepage(ctx, version, config)
}

func (service *PublicConfigService) composeHomepage(
	ctx context.Context,
	version uint64,
	config model.HomepageConfig,
) (PublicHomepage, error) {
	var socialLinks []PublicSocialLink
	modules := make([]PublicHomepageModule, 0, len(config.Modules))
	for _, module := range config.Modules {
		if !module.Enabled {
			continue
		}

		publicModule, include, err := service.composeHomepageModule(ctx, module, &socialLinks)
		if err != nil {
			return PublicHomepage{}, err
		}
		if include {
			modules = append(modules, publicModule)
		}
	}

	sort.SliceStable(modules, func(left, right int) bool {
		if modules[left].SortOrder != modules[right].SortOrder {
			return modules[left].SortOrder < modules[right].SortOrder
		}
		return homepageModuleRank(modules[left].Type) < homepageModuleRank(modules[right].Type)
	})

	return PublicHomepage{Version: version, Modules: modules}, nil
}

func (service *PublicConfigService) siteSetting(ctx context.Context) (model.SiteSetting, error) {
	var setting model.SiteSetting
	result := service.database.WithContext(ctx).Where("id = ?", uint64(1)).Limit(1).Find(&setting)
	if result.Error != nil {
		return model.SiteSetting{}, fmt.Errorf("find public site setting: %w", result.Error)
	}
	if result.RowsAffected == 0 {
		return model.DefaultSiteSetting(time.Now()), nil
	}
	if err := model.ValidateSiteSetting(setting); err != nil {
		slog.Warn("public site setting is invalid; using defaults", "error", err)
		return model.DefaultSiteSetting(time.Now()), nil
	}
	return setting, nil
}

func (service *PublicConfigService) publishedHomepageConfig(
	ctx context.Context,
) (uint64, model.HomepageConfig, error) {
	var version model.HomepageVersion
	result := service.database.WithContext(ctx).
		Where("status = ?", model.HomepageStatusPublished).
		Limit(1).
		Find(&version)
	if result.Error != nil {
		return 0, model.HomepageConfig{}, fmt.Errorf("find published homepage config: %w", result.Error)
	}
	if result.RowsAffected == 0 {
		return 0, model.DefaultHomepageConfig(), nil
	}

	config, err := model.DecodeHomepageConfig(version.ConfigJSON)
	if err != nil {
		slog.Warn("published homepage config is invalid; using defaults", "version", version.VersionNumber, "error", err)
		return 0, model.DefaultHomepageConfig(), nil
	}
	return version.VersionNumber, config, nil
}

func (service *PublicConfigService) composeHomepageModule(
	ctx context.Context,
	module model.HomepageModule,
	cachedSocialLinks *[]PublicSocialLink,
) (PublicHomepageModule, bool, error) {
	publicModule := PublicHomepageModule{Type: module.Type, SortOrder: module.SortOrder}
	switch module.Type {
	case model.HomepageModuleHero:
		config := *module.Hero
		publicModule.Hero = &config
	case model.HomepageModuleAbout:
		config := *module.About
		publicModule.About = &config
	case model.HomepageModuleFeaturedArticles:
		articles, err := service.featuredArticles(ctx, module.FeaturedArticles.Limit)
		if err != nil {
			return PublicHomepageModule{}, false, err
		}
		if len(articles) == 0 {
			return PublicHomepageModule{}, false, nil
		}
		publicModule.FeaturedArticles = &PublicFeaturedArticlesModuleConfig{
			Title:       module.FeaturedArticles.Title,
			Description: module.FeaturedArticles.Description,
			Limit:       module.FeaturedArticles.Limit,
			Articles:    articles,
		}
	case model.HomepageModuleLatestArticles:
		page, err := service.articleService.ListPublished(ctx, 1, module.LatestArticles.Limit)
		if err != nil {
			return PublicHomepageModule{}, false, fmt.Errorf("compose latest articles module: %w", err)
		}
		publicModule.LatestArticles = &PublicLatestArticlesModuleConfig{
			Title:            module.LatestArticles.Title,
			Description:      module.LatestArticles.Description,
			Limit:            module.LatestArticles.Limit,
			ShowSummary:      module.LatestArticles.ShowSummary,
			ShowDate:         module.LatestArticles.ShowDate,
			ShowCommentCount: module.LatestArticles.ShowCommentCount,
			ShowViewAll:      module.LatestArticles.ShowViewAll,
			Articles:         publicArticleSummaries(page.Items),
		}
	case model.HomepageModuleTechStack:
		items := make([]model.TechItem, 0, len(module.TechStack.Items))
		for _, item := range module.TechStack.Items {
			if item.IsVisible {
				items = append(items, item)
			}
		}
		sort.SliceStable(items, func(left, right int) bool {
			if items[left].SortOrder != items[right].SortOrder {
				return items[left].SortOrder < items[right].SortOrder
			}
			return items[left].Name < items[right].Name
		})
		publicItems := make([]PublicTechItem, 0, len(items))
		for _, item := range items {
			publicItems = append(publicItems, PublicTechItem{
				Name: item.Name, Description: item.Description, IconURL: item.IconURL, URL: item.URL,
			})
		}
		publicModule.TechStack = &PublicTechStackModuleConfig{
			Title: module.TechStack.Title, Description: module.TechStack.Description, Items: publicItems,
		}
	case model.HomepageModuleSocialLinks:
		if *cachedSocialLinks == nil {
			links, err := service.publicSocialLinks(ctx)
			if err != nil {
				return PublicHomepageModule{}, false, err
			}
			*cachedSocialLinks = links
		}
		publicModule.SocialLinks = &PublicSocialLinksModuleConfig{
			Title: module.SocialLinks.Title, Description: module.SocialLinks.Description, Links: *cachedSocialLinks,
		}
	default:
		slog.Warn("skipping unsupported public homepage module", "type", module.Type)
		return PublicHomepageModule{}, false, nil
	}
	return publicModule, true, nil
}

func (service *PublicConfigService) featuredArticles(
	ctx context.Context,
	limit int,
) ([]PublicArticleSummary, error) {
	articles := make([]PublicArticleSummary, 0)
	result := service.database.WithContext(ctx).
		Table("featured_articles").
		Select(
			"articles.id, articles.title, articles.slug, articles.summary, articles.published_at, "+
				"(SELECT COUNT(*) FROM comments WHERE comments.article_id = articles.id "+
				"AND comments.status = ?) AS comment_count",
			model.CommentStatusApproved,
		).
		Joins("JOIN articles ON articles.id = featured_articles.article_id").
		Where("featured_articles.is_visible = ? AND articles.status = ?", true, model.ArticleStatusPublished).
		Order("featured_articles.sort_order ASC, featured_articles.id ASC").
		Limit(limit).
		Scan(&articles)
	if result.Error != nil {
		return nil, fmt.Errorf("list public featured articles: %w", result.Error)
	}
	return articles, nil
}

func (service *PublicConfigService) publicSocialLinks(ctx context.Context) ([]PublicSocialLink, error) {
	var links []model.SocialLink
	if err := service.database.WithContext(ctx).
		Where("is_visible = ?", true).
		Order("sort_order ASC, id ASC").
		Find(&links).Error; err != nil {
		return nil, fmt.Errorf("list homepage social links: %w", err)
	}
	return validPublicSocialLinks(links), nil
}

func (module PublicHomepageModule) MarshalJSON() ([]byte, error) {
	var config any
	switch module.Type {
	case model.HomepageModuleHero:
		config = module.Hero
	case model.HomepageModuleAbout:
		config = module.About
	case model.HomepageModuleFeaturedArticles:
		config = module.FeaturedArticles
	case model.HomepageModuleLatestArticles:
		config = module.LatestArticles
	case model.HomepageModuleTechStack:
		config = module.TechStack
	case model.HomepageModuleSocialLinks:
		config = module.SocialLinks
	default:
		return nil, fmt.Errorf("unsupported public homepage module type %q", module.Type)
	}
	if config == nil {
		return nil, errors.New("public homepage module config is missing")
	}
	return json.Marshal(struct {
		Type      string `json:"type"`
		SortOrder int    `json:"sort_order"`
		Config    any    `json:"config"`
	}{Type: module.Type, SortOrder: module.SortOrder, Config: config})
}

func publicSiteConfig(
	setting model.SiteSetting,
	navigation []PublicNavigationItem,
	socialLinks []PublicSocialLink,
) PublicSiteConfig {
	return PublicSiteConfig{
		Site: PublicSite{
			Name: setting.SiteName, ShortName: stringValue(setting.SiteShortName), Description: setting.SiteDescription,
			TitleSuffix: stringValue(setting.TitleSuffix), LogoURL: stringValue(setting.LogoURL),
			FaviconURL: stringValue(setting.FaviconURL), DefaultShareImageURL: stringValue(setting.DefaultShareImageURL),
		},
		Navigation:  navigation,
		SocialLinks: socialLinks,
		Footer: PublicFooter{
			CopyrightName: setting.CopyrightName, StartYear: setting.StartYear,
			AdditionalText: stringValue(setting.AdditionalText), FilingNumber: stringValue(setting.FilingNumber),
			FilingURL: stringValue(setting.FilingURL), ShowTechnology: setting.ShowTechnology,
			TechnologyText: stringValue(setting.TechnologyText),
		},
	}
}

func validPublicNavigation(items []model.NavigationItem) []PublicNavigationItem {
	result := make([]PublicNavigationItem, 0, len(items))
	for _, item := range items {
		if len(result) == 10 {
			break
		}
		if err := model.ValidateNavigationItems([]model.NavigationItem{item}); err != nil {
			slog.Warn("skipping invalid public navigation item", "id", item.ID, "error", err)
			continue
		}
		result = append(result, PublicNavigationItem{
			Name: item.Name, URL: item.URL, LinkType: item.LinkType, OpenInNewTab: item.OpenInNewTab,
		})
	}
	return result
}

func validPublicSocialLinks(links []model.SocialLink) []PublicSocialLink {
	result := make([]PublicSocialLink, 0, len(links))
	for _, link := range links {
		if len(result) == 15 {
			break
		}
		if err := model.ValidateSocialLinks([]model.SocialLink{link}); err != nil {
			slog.Warn("skipping invalid public social link", "id", link.ID, "error", err)
			continue
		}
		result = append(result, PublicSocialLink{
			Platform: link.Platform, DisplayName: link.DisplayName, URL: link.URL,
		})
	}
	return result
}

func publicArticleSummaries(articles []model.Article) []PublicArticleSummary {
	result := make([]PublicArticleSummary, 0, len(articles))
	for _, article := range articles {
		result = append(result, PublicArticleSummary{
			ID: article.ID, Title: article.Title, Slug: article.Slug, Summary: article.Summary,
			PublishedAt: article.PublishedAt, CommentCount: article.CommentCount,
		})
	}
	return result
}

func homepageModuleRank(moduleType string) int {
	switch moduleType {
	case model.HomepageModuleHero:
		return 0
	case model.HomepageModuleAbout:
		return 1
	case model.HomepageModuleFeaturedArticles:
		return 2
	case model.HomepageModuleLatestArticles:
		return 3
	case model.HomepageModuleTechStack:
		return 4
	case model.HomepageModuleSocialLinks:
		return 5
	default:
		return 6
	}
}

func stringValue(value *string) string {
	if value == nil {
		return ""
	}
	return *value
}

File: frontend/src/utils/publicConfig.ts
import type {
  ArticleSummary,
  PublicAboutConfig,
  PublicArticlesModuleConfig,
  PublicHeroConfig,
  PublicHomepageButton,
  PublicHomepageConfig,
  PublicHomepageModule,
  PublicHomepageModuleType,
  PublicLatestArticlesConfig,
  PublicNavigationItem,
  PublicSiteConfig,
  PublicSocialLink,
  PublicSocialLinksConfig,
  PublicTechItem,
  PublicTechStackConfig,
} from '../types/blog'

export const DEFAULT_PUBLIC_SITE_CONFIG: PublicSiteConfig = {
  site: {
    name: '字里行间',
    short_name: '字里行间',
    description: '记录开发、阅读与生活中的思考。',
    title_suffix: '字里行间',
    logo_url: '',
    favicon_url: '',
    default_share_image_url: '',
  },
  navigation: [
    { name: '首页', url: '/', link_type: 'internal', open_in_new_tab: false },
    { name: '归档', url: '/archive', link_type: 'internal', open_in_new_tab: false },
    { name: '关于', url: '/about', link_type: 'internal', open_in_new_tab: false },
  ],
  social_links: [],
  footer: {
    copyright_name: '字里行间',
    start_year: new Date().getFullYear(),
    additional_text: '',
    filing_number: '',
    filing_url: '',
    show_technology: true,
    technology_text: 'Built with Vue 3 and Go',
  },
}

const knownModuleTypes = new Set<PublicHomepageModuleType>([
  'hero', 'about', 'featured_articles', 'latest_articles', 'tech_stack', 'social_links',
])

function record(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null
}

function text(value: unknown, fallback = '', maximum = 2000): string {
  return typeof value === 'string' ? Array.from(value.trim()).slice(0, maximum).join('') : fallback
}

function bool(value: unknown, fallback = false): boolean {
  return typeof value === 'boolean' ? value : fallback
}

function integer(value: unknown, fallback: number, minimum: number, maximum: number): number {
  return typeof value === 'number' && Number.isInteger(value) && value >= minimum && value <= maximum
    ? value
    : fallback
}

function httpURL(value: unknown): string {
  const candidate = text(value, '', 500)
  if (!candidate) return ''
  try {
    const parsed = new URL(candidate)
    return parsed.protocol === 'http:' || parsed.protocol === 'https:' ? candidate : ''
  } catch {
    return ''
  }
}

function internalURL(value: unknown): string {
  const candidate = text(value, '', 500)
  return candidate.startsWith('/') && !candidate.startsWith('//') && !candidate.includes('\n') && !candidate.includes('\r')
    ? candidate
    : ''
}

function publicLink(value: unknown, allowMailto = false): string {
  const candidate = text(value, '', 500)
  if (allowMailto && /^mailto:[^\s@]+@[^\s@]+\.[^\s@]+$/i.test(candidate)) return candidate
  return httpURL(candidate)
}

function normalizeNavigation(value: unknown): PublicNavigationItem[] {
  if (!Array.isArray(value)) return DEFAULT_PUBLIC_SITE_CONFIG.navigation.map((item) => ({ ...item }))
  return value.slice(0, 10).flatMap((raw) => {
    const item = record(raw)
    if (!item) return []
    const linkType = item.link_type === 'external' ? 'external' : 'internal'
    const url = linkType === 'external' ? httpURL(item.url) : internalURL(item.url)
    const name = text(item.name, '', 20)
    return name && url ? [{ name, url, link_type: linkType, open_in_new_tab: bool(item.open_in_new_tab) }] : []
  })
}

function normalizeSocialLinks(value: unknown): PublicSocialLink[] {
  if (!Array.isArray(value)) return []
  return value.slice(0, 15).flatMap((raw) => {
    const item = record(raw)
    if (!item) return []
    const platform = text(item.platform, 'custom', 30)
    const displayName = text(item.display_name, '', 30)
    const url = publicLink(item.url, platform === 'email')
    return displayName && url ? [{ platform, display_name: displayName, url }] : []
  })
}

export function normalizePublicSiteConfig(value: unknown): PublicSiteConfig {
  const input = record(value)
  if (!input) return structuredClone(DEFAULT_PUBLIC_SITE_CONFIG)
  const site = record(input.site) ?? {}
  const footer = record(input.footer) ?? {}
  const name = text(site.name, DEFAULT_PUBLIC_SITE_CONFIG.site.name, 50) || DEFAULT_PUBLIC_SITE_CONFIG.site.name
  return {
    site: {
      name,
      short_name: text(site.short_name, name, 20) || name,
      description: text(site.description, DEFAULT_PUBLIC_SITE_CONFIG.site.description, 200) || DEFAULT_PUBLIC_SITE_CONFIG.site.description,
      title_suffix: text(site.title_suffix, name, 50),
      logo_url: httpURL(site.logo_url),
      favicon_url: httpURL(site.favicon_url),
      default_share_image_url: httpURL(site.default_share_image_url),
    },
    navigation: normalizeNavigation(input.navigation),
    social_links: normalizeSocialLinks(input.social_links),
    footer: {
      copyright_name: text(footer.copyright_name, name, 50) || name,
      start_year: typeof footer.start_year === 'number' && footer.start_year >= 1900 && footer.start_year <= new Date().getFullYear()
        ? Math.trunc(footer.start_year)
        : null,
      additional_text: text(footer.additional_text, '', 200),
      filing_number: text(footer.filing_number, '', 100),
      filing_url: httpURL(footer.filing_url),
      show_technology: bool(footer.show_technology, true),
      technology_text: text(footer.technology_text, '', 100),
    },
  }
}

function normalizeButton(value: unknown, fallback: PublicHomepageButton): PublicHomepageButton {
  const input = record(value) ?? {}
  const linkType = input.link_type === 'external' ? 'external' : 'internal'
  const url = linkType === 'external' ? httpURL(input.url) : internalURL(input.url)
  return {
    enabled: bool(input.enabled, fallback.enabled) && Boolean(url),
    text: text(input.text, fallback.text, 20) || fallback.text,
    url: url || fallback.url,
    link_type: linkType,
    open_in_new_tab: bool(input.open_in_new_tab, fallback.open_in_new_tab),
  }
}

function normalizeHero(value: unknown): PublicHeroConfig {
  const input = record(value) ?? {}
  return {
    eyebrow: text(input.eyebrow, 'WELCOME', 50),
    title: text(input.title, '在字里行间，保存思想的回声。', 100) || '在字里行间，保存思想的回声。',
    highlight_text: text(input.highlight_text, '', 50),
    description: text(input.description, '记录开发、阅读与生活中的思考。', 300) || '记录开发、阅读与生活中的思考。',
    image_url: httpURL(input.image_url),
    background_image_url: httpURL(input.background_image_url),
    layout: input.layout === 'left' ? 'left' : 'center',
    primary_button: normalizeButton(input.primary_button, {
      enabled: true, text: '开始阅读', url: '/archive', link_type: 'internal', open_in_new_tab: false,
    }),
    secondary_button: normalizeButton(input.secondary_button, {
      enabled: true, text: '关于本站', url: '/about', link_type: 'internal', open_in_new_tab: false,
    }),
  }
}

function normalizeAbout(value: unknown): PublicAboutConfig {
  const input = record(value) ?? {}
  const position = input.image_position === 'left' || input.image_position === 'right' ? input.image_position : 'none'
  return {
    title: text(input.title, '关于我', 100) || '关于我',
    description: text(input.description, '', 200),
    content: text(input.content, '这里记录开发、阅读与生活中的思考。', 2000) || '这里记录开发、阅读与生活中的思考。',
    image_url: httpURL(input.image_url),
    image_position: position,
  }
}

function normalizeArticle(value: unknown): ArticleSummary | null {
  const input = record(value)
  if (!input) return null
  const id = integer(input.id, 0, 1, Number.MAX_SAFE_INTEGER)
  const title = text(input.title, '', 200)
  const slug = text(input.slug, '', 200)
  const publishedAt = text(input.published_at, '', 100)
  if (!id || !title || !slug || !publishedAt) return null
  return {
    id,
    title,
    slug,
    summary: text(input.summary, '', 1000),
    published_at: publishedAt,
    comment_count: integer(input.comment_count, 0, 0, Number.MAX_SAFE_INTEGER),
  }
}

function normalizeArticles(value: unknown): ArticleSummary[] {
  return Array.isArray(value) ? value.map(normalizeArticle).filter((item): item is ArticleSummary => Boolean(item)) : []
}

function normalizeFeatured(value: unknown): PublicArticlesModuleConfig {
  const input = record(value) ?? {}
  return {
    title: text(input.title, '推荐文章', 100) || '推荐文章',
    description: text(input.description, '', 200),
    limit: integer(input.limit, 3, 1, 10),
    articles: normalizeArticles(input.articles),
  }
}

function normalizeLatest(value: unknown): PublicLatestArticlesConfig {
  const input = record(value) ?? {}
  return {
    title: text(input.title, '最新文章', 100) || '最新文章',
    description: text(input.description, '', 200),
    limit: integer(input.limit, 10, 3, 20),
    show_summary: bool(input.show_summary, true),
    show_date: bool(input.show_date, true),
    show_comment_count: bool(input.show_comment_count, true),
    show_view_all: bool(input.show_view_all, true),
    articles: normalizeArticles(input.articles),
  }
}

function normalizeTech(value: unknown): PublicTechStackConfig {
  const input = record(value) ?? {}
  const items = Array.isArray(input.items) ? input.items.slice(0, 20).flatMap((raw): PublicTechItem[] => {
    const item = record(raw)
    const name = text(item?.name, '', 30)
    return name ? [{ name, description: text(item?.description, '', 100), icon_url: httpURL(item?.icon_url), url: httpURL(item?.url) }] : []
  }) : []
  return {
    title: text(input.title, '技术栈', 100) || '技术栈',
    description: text(input.description, '', 200),
    items,
  }
}

function normalizeSocialModule(value: unknown): PublicSocialLinksConfig {
  const input = record(value) ?? {}
  return {
    title: text(input.title, '找到我', 100) || '找到我',
    description: text(input.description, '', 200),
    links: normalizeSocialLinks(input.links),
  }
}

export function normalizePublicHomepage(value: unknown): PublicHomepageConfig {
  const input = record(value)
  const rawModules = Array.isArray(input?.modules) ? input.modules : []
  const modules: PublicHomepageModule[] = []
  const seen = new Set<PublicHomepageModuleType>()
  for (const raw of rawModules) {
    const module = record(raw)
    if (!module || typeof module.type !== 'string' || !knownModuleTypes.has(module.type as PublicHomepageModuleType)) {
      console.warn('Skipping unsupported homepage module', module?.type)
      continue
    }
    const type = module.type as PublicHomepageModuleType
    if (seen.has(type)) continue
    seen.add(type)
    const sortOrder = integer(module.sort_order, modules.length * 10 + 10, -100000, 100000)
    if (type === 'hero') modules.push({ type, sort_order: sortOrder, config: normalizeHero(module.config) })
    if (type === 'about') modules.push({ type, sort_order: sortOrder, config: normalizeAbout(module.config) })
    if (type === 'featured_articles') {
      const config = normalizeFeatured(module.config)
      if (config.articles.length) modules.push({ type, sort_order: sortOrder, config })
    }
    if (type === 'latest_articles') modules.push({ type, sort_order: sortOrder, config: normalizeLatest(module.config) })
    if (type === 'tech_stack') {
      const config = normalizeTech(module.config)
      if (config.items.length) modules.push({ type, sort_order: sortOrder, config })
    }
    if (type === 'social_links') {
      const config = normalizeSocialModule(module.config)
      if (config.links.length) modules.push({ type, sort_order: sortOrder, config })
    }
  }
  modules.sort((left, right) => left.sort_order - right.sort_order)
  return { version: integer(input?.version, 0, 0, Number.MAX_SAFE_INTEGER), modules }
}
